### DISCUSSION TITLE
Why use reflect to access RedisProperties method?

### DISCUSSION THREAD

**Discussion by Kicey** [redisson/redisson#4838] [Q&A] [answered]
Method timeoutMethod = ReflectionUtils.findMethod(RedisProperties.class, "getTimeout");
        Object timeoutValue = ReflectionUtils.invokeMethod(timeoutMethod, redisProperties);
        int timeout;
Above is the code in the `org.redisson.spring.starter.RedissonAutoConfiguration`.
I'm reading the auto configuration part of redisson spring starter, I'm curious about why using reflect to access the RedisProperties method instead of accessing it directly?

**Comment by Milk-Bottle** [comment, marked as answer]
**The answer comes from chatGPT**

The reason why reflection is used in RedissonAutoConfiguration to access the RedisProperties method instead of accessing it directly is because RedisProperties is an optional dependency in Redisson Spring Boot starter.

If RedisProperties is not present on the classpath, attempting to access it directly would result in a ClassNotFoundException. To avoid this issue, reflection is used to check whether RedisProperties is present, and only access it through reflection if it is available.

Here's the relevant code from RedissonAutoConfiguration:

`
    private void configureSingleServerConfig(SingleServerConfig singleServerConfig, RedisProperties properties) throws Exception {
        Class<?> redisPropertiesClass = RedisProperties.class;
        if (ClassUtils.isPresent(redisPropertiesClass.getName(), getClass().getClassLoader())) {
            Method method = redisPropertiesClass.getMethod("getDatabase");
            Integer database = (Integer) ReflectionUtils.invokeMethod(method, properties);
            if (database != null) {
                singleServerConfig.setDatabase(database);
            }
            method = redisPropertiesClass.getMethod("getUrl");
            String url = (String) ReflectionUtils.invokeMethod(method, properties);
            if (StringUtils.hasText(url)) {
                RedisURI redisUri = RedisURI.create(url);
                singleServerConfig.setAddress(redisUri.toSocketAddress());
                if (redisUri.getPassword() != null) {
                    singleServerConfig.setPassword(redisUri.getPassword());
                }
            }
            method = redisPropertiesClass.getMethod("getHost");
            String host = (String) ReflectionUtils.invokeMethod(method, properties);
            if (StringUtils.hasText(host)) {
                singleServerConfig.setAddress(host + ":" + properties.getPort());
                if (StringUtils.hasText(properties.getPassword())) {
                    singleServerConfig.setPassword(properties.getPassword());
                }
            }
        ...
        }
    ...
    }

`

As you can see, ClassUtils.isPresent() is used to check if RedisProperties is present, and then ReflectionUtils.invokeMethod() is used to invoke the relevant getter methods on the RedisProperties object. This approach allows the RedissonAutoConfiguration to gracefully handle the absence of RedisProperties on the classpath.

**Reply by mrniko** [reply]
Hi is smart 👍
